web/packages/teleterm/src/ui/services/clusters/clustersService.ts

```diff
@@ -640,3 +640,24 @@ export function makeApp(source: tsh.App): App {
 
   return { ...source, addrWithProtocol };
 }
+
+/** Returns a URL that can be used to open the app in a browser. */
+export function getWebAppLaunchUrl(
```

Comment:
It makes this function a bit more difficult to use (since we have to get cluster and root cluster in callsites), but definitely it is a better API. We just need an easier way to get root/leaf cluster from a resource URI.